File: messagebus/src/tests/simpleprotocol/simpleprotocol.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include <vespa/messagebus/testlib/simpleprotocol.h>
#include <vespa/messagebus/testlib/simplemessage.h>
#include <vespa/messagebus/testlib/simplereply.h>
#include <vespa/messagebus/testlib/slobrok.h>
#include <vespa/messagebus/testlib/testserver.h>
#include <vespa/messagebus/ireplyhandler.h>
#include <vespa/messagebus/routing/routingcontext.h>
#include <vespa/vespalib/gtest/gtest.h>
#include <vespa/vespalib/component/vtag.h>

using namespace mbus;

TEST(SimpleProtocolTest, simpleprotocol_test) {

    vespalib::Version version = vespalib::Vtag::currentVersion;
    SimpleProtocol protocol;
    EXPECT_TRUE(protocol.getName() == "Simple");

    EXPECT_EQ(24u + 2 *sizeof(std::string), sizeof(Result));
    EXPECT_EQ(8u + 2 *sizeof(std::string), sizeof(Error));
    EXPECT_EQ(56u, sizeof(Routable));
    {
        // test protocol
        IRoutingPolicy::UP bogus = protocol.createPolicy("bogus", "");
        EXPECT_FALSE(bogus);
    }
    {
        // test SimpleMessage
        EXPECT_EQ(104u, sizeof(Message));
        EXPECT_EQ(120u + sizeof(std::string), sizeof(SimpleMessage));
        auto msg = std::make_unique<SimpleMessage>("test");
        EXPECT_TRUE(!msg->isReply());
        EXPECT_TRUE(msg->getProtocol() == SimpleProtocol::NAME);
        EXPECT_TRUE(msg->getType() == SimpleProtocol::MESSAGE);
        EXPECT_TRUE(static_cast<SimpleMessage&>(*msg).getValue() == "test");
        Blob b = protocol.encode(version, *msg);
        EXPECT_TRUE(b.size() > 0);
        Routable::UP tmp = protocol.decode(version, BlobRef(b));
        ASSERT_TRUE(tmp);
        EXPECT_TRUE(!tmp->isReply());
        EXPECT_TRUE(tmp->getProtocol() == SimpleProtocol::NAME);
        EXPECT_TRUE(tmp->getType() == SimpleProtocol::MESSAGE);
        EXPECT_TRUE(static_cast<SimpleMessage&>(*tmp).getValue() == "test");
    }
    {
        // test SimpleReply
        EXPECT_EQ(96u, sizeof(Reply));
        EXPECT_EQ(96u + sizeof(std::string), sizeof(SimpleReply));
        auto reply = std::make_unique<SimpleReply>("reply");
        EXPECT_TRUE(reply->isReply());
        EXPECT_TRUE(reply->getProtocol() == SimpleProtocol::NAME);
        EXPECT_TRUE(reply->getType() == SimpleProtocol::REPLY);
        EXPECT_TRUE(static_cast<SimpleReply&>(*reply).getValue() == "reply");
        Blob b = protocol.encode(version, *reply);
        EXPECT_TRUE(b.size() > 0);
        Routable::UP tmp = protocol.decode(version, BlobRef(b));
        ASSERT_TRUE(tmp);
        EXPECT_TRUE(tmp->isReply());
        EXPECT_TRUE(tmp->getProtocol() == SimpleProtocol::NAME);
        EXPECT_TRUE(tmp->getType() == SimpleProtocol::REPLY);
        EXPECT_TRUE(static_cast<SimpleReply&>(*tmp).getValue() == "reply");
    }
}

GTEST_MAIN_RUN_ALL_TESTS()

File: messagebus/src/main/java/com/yahoo/messagebus/test/SimpleProtocol.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.messagebus.test;

import com.yahoo.component.Version;
import com.yahoo.messagebus.Protocol;
import com.yahoo.messagebus.Routable;
import com.yahoo.messagebus.routing.RoutingPolicy;
import com.yahoo.text.Utf8;
import com.yahoo.text.Utf8String;

import java.util.HashMap;
import java.util.Map;

/**
 * @author Simon Thoresen Hult
 */
public class SimpleProtocol implements Protocol {

    public static final Utf8String NAME = new Utf8String("Simple");
    public static final int MESSAGE = 1;
    public static final int REPLY = 2;
    private final Map<String, PolicyFactory> policies = new HashMap<String, PolicyFactory>();

    @Override
    public String getName() {
        return NAME.toString();
    }

    @Override
    public RoutingPolicy createPolicy(String name, String param) {
        if (policies.containsKey(name)) {
            return policies.get(name).create(param);
        }
        return null;
    }

    @Override
    public Routable decode(Version version, byte[] data) {
        String str = Utf8.toString(data);
        if (str.length() < 1) {
            return null;
        }
        char c = str.charAt(0);
        if (c == 'M') {
            return new SimpleMessage(str.substring(1));
        }
        if (c == 'R') {
            return new SimpleReply(str.substring(1));
        }
        return null;
    }

    @Override
    public byte[] encode(Version version, Routable routable) {
        if (routable.getType() == MESSAGE) {
            return Utf8.toBytes("M" + ((SimpleMessage)routable).getValue());
        } else if (routable.getType() == REPLY) {
            return Utf8.toBytes("R" + ((SimpleReply)routable).getValue());
        } else {
            return null;
        }
    }

    /**
     * Registers a policy factory with this protocol under a given name. Whenever a policy is requested that matches
     * this name, the factory is invoked.
     *
     * @param name    The name of the policy.
     * @param factory The policy factory.
     */
    public void addPolicyFactory(String name, PolicyFactory factory) {
        policies.put(name, factory);
    }

    /**
     * Defines a policy factory interface that tests can use to register arbitrary policies with this protocol.
     */
    public interface PolicyFactory {

        /**
         * Creates a new instance of the routing policy that this factory encapsulates.
         *
         * @param param The param for the policy constructor.
         * @return The routing policy created.
         */
        public RoutingPolicy create(String param);
    }
}
